From b353f194952510df046fbd3e24185aa78eb2717f Mon Sep 17 00:00:00 2001
From: Xulin Sun <xulin.sun@windriver.com>
Date: Tue, 5 Mar 2019 16:56:15 +0800
Subject: [PATCH 8/8] Revert "arm64: Allocate elfcorehdr & crashkernel mem
 before any reservation"

This reverts commit 363ca7d2737be42225196a460bb243e7edaa3250.
To avoid below calltrace when the second kernel of kdump is run.

Unable to handle kernel write to read-only memory at virtual address ffff800020080000
Mem abort info:
  ESR = 0x9600004f
  Exception class = DABT (current EL), IL = 32 bits
  SET = 0, FnV = 0
  EA = 0, S1PTW = 0
Data abort info:
  ISV = 0, ISS = 0x0000004f
  CM = 0, WnR = 1
swapper pgtable: 4k pages, 48-bit VAs, pgdp = (____ptrval____)
[ffff800020080000] pgd=00000009ffff7803, pud=00000009ffff6803, pmd=00000009ffff5803, pte=00e00009e0080f93
Internal error: Oops: 9600004f [#1] PREEMPT SMP
Modules linked in:
CPU: 0 PID: 1 Comm: swapper/0 Not tainted 4.18.20-rt8-yocto-preempt-rt #1
Hardware name: LS1046A RDB Board (DT)
pstate: 20000005 (nzCv daif -PAN -UAO)
pc : __memset+0x16c/0x188
lr : qm_reserve_memory+0xd4/0xfc
sp : ffff00000803bc10
x29: ffff00000803bc10 x28: ffff000008ea0564
x27: ffff000008fa1018 x26: ffff000008f49040
x25: ffff0000091a3000 x24: ffff0000091a3b50
x23: ffff000009290000 x22: 0000000000000000
x21: 0000000000000001 x20: 0000000000800000
x19: ffff800020000000 x18: 0000000000000030
x17: 0000000080000000 x16: 0000000000000000
x15: ffffffffffffffff x14: 0000000000000000
x13: 000000000000002d x12: 00000009ffffffd0
x11: 0000000000000000 x10: 0000000000000031
x9 : 0000000000000000 x8 : ffff800020080000
x7 : 0000000000000000 x6 : 000000000000003f
x5 : 0000000000000040 x4 : 0000000000000000
x3 : 0000000000000004 x2 : 000000000077ffc0
x1 : 0000000000000000 x0 : ffff800020000000
Process swapper/0 (pid: 1, stack limit = 0x(____ptrval____))
Call trace:
 __memset+0x16c/0x188
 fsl_qman_init+0x1c0/0x358
 qman_init_early+0x74/0xa8
 do_one_initcall+0x68/0x270
 kernel_init_freeable+0x32c/0x424
 kernel_init+0x18/0x108
 ret_from_fork+0x10/0x1c
Code: 91010108 54ffff4a 8b040108 cb050042 (d50b7428)

Signed-off-by: Xulin Sun <xulin.sun@windriver.com>
---
 arch/arm64/mm/init.c |   12 ++++--------
 1 file changed, 4 insertions(+), 8 deletions(-)

diff --git a/arch/arm64/mm/init.c b/arch/arm64/mm/init.c
index f589f4a..787e279 100644
--- a/arch/arm64/mm/init.c
+++ b/arch/arm64/mm/init.c
@@ -460,14 +460,6 @@ void __init arm64_memblock_init(void)
 	 * Register the kernel text, kernel data, initrd, and initial
 	 * pagetables with memblock.
 	 */
-
-	/* make this the first reservation so that there are no chances of
-	 * overlap
-	 */
-	reserve_elfcorehdr();
-
-	reserve_crashkernel();
-
 	memblock_reserve(__pa_symbol(_text), _end - _text);
 #ifdef CONFIG_BLK_DEV_INITRD
 	if (initrd_start) {
@@ -487,6 +479,10 @@ void __init arm64_memblock_init(void)
 	else
 		arm64_dma_phys_limit = PHYS_MASK + 1;
 
+	reserve_crashkernel();
+
+	reserve_elfcorehdr();
+
 	high_memory = __va(memblock_end_of_DRAM() - 1) + 1;
 
 	dma_contiguous_reserve(arm64_dma_phys_limit);
-- 
1.7.9.5

